### ISSUE TITLE
More aggressive backoff on rate throttling

### ISSUE THREAD

**Issue by prashanth26** [gardener/gardener#3120]
**How to categorize this issue?**

/area ops-productivity
/kind enhancement
/kind discussion
/priority normal

**What would you like to be added**:
Currently, shoot reconciliations/deletions on infrastructure failures are retried too often. Hence we would like this backoff to be more aggressive to allow at least a few shoot reconciliations to succeed without rate limits and avoid all shoots to be reconciling together. 

Below shows one such error message. 
```
Waiting until shoot infrastructure has been destroyed" failed: Failed to delete Infrastructure shoot--xxx--yyy--zzz Error deleting infrastructure: 1 error occurred:
	* Failed to destroy load balancers and security groups: Throttling: Rate exceeded
	status code: 400, request id: xxx-yyy-zzz
```

**Why is this needed**:
This especially causes issues when several clusters are using the same credentials and their rate limits are hit. Although there exists a backoff mechanism it takes a long time for the shoots to be marked as failed to avoid further reconciliation and allow healing of clusters after a long time.

**Comment by timebertt** [comment]
Thanks for opening this issue.

> shoot reconciliations/deletions on infrastructure failures are retried too often

As you also pointed out, this issue start to occur when many clusters use the same IaaS credentials (and by that all ccm's and other controllers contribute to the same rate limit budget).
Do you think, this will really be improved by more aggressive backoff in the infrastructure deletion?
The infra reconciliation/deletion only contributes a part to the total number of IaaS API calls, so by more aggressive backoff the situation might not improve at all and we might even miss a chance to retry the operation at a "more quite" time.

So in short, I understand your proposal. But I'm not sure, if we can even do something here from gardener's perspective. 
Probably, the best solution will be to use different credentials (if this extends the rate limit budget?).

**Comment by prashanth26** [comment]
>by that all ccm's and other controllers contribute to the same rate limit budget

Oh yes, that is also a valid point that skipped my mind. In that case, there is probably not much we can do here from Gardener perspective. 

>Probably, the best solution will be to use different credentials (if this extends the rate limit budget?).

This is something I have already proposed to the shoot owner running into this issue. However, they mentioned that their architecture will need tweaking to change this as they have a single credential used in their architecture I suspect.

**Comment by timuthy** [comment]
A document explaining the situation and recommending a best practice can also be a form of mitigation, especially for new users.

**Comment by vpnachev** [comment]
While Gardener is contributing with just few API calls during infrastructure deletion, it still cannot delete the resources. 
In this case I think it would be better to set the shoot as failed right away and prevent further reconciliations.

**Comment by BeckerMax** [comment]
Mitigations:
1. Make error transparent in the shoot status (currently just shown in the infrastructure logs and not shown in the shoot status)
2. Inform endusers about using different infrastructure accounts (consequences if you use the same infrastructure account for multiple Shoot clusters).
3. Use more aggressive backoffs in case of rate limit errors

In the quality week we will start with issue 1 and 2. The third one still needs clarification.

**Comment by BeckerMax** [comment]
Regarding 1.:
Actually, the errors are already available in the Shoot Status. In question was also a context deadline exceeded error:
```
task "Waiting until shoot infrastructure has been destroyed" failed: Failed to delete Infrastructure shoot--xxx: Error deleting infrastructure: 1 error occurred:
	* Failed to destroy load balancers and security groups: RequestCanceled: request context canceled
caused by: context deadline exceeded
```
I checked it and the `RequestCanceled` error is exactly the one we receive from the aws-go-sdk and correctly propagate to the user. See:
https://github.com/gardener/gardener-extension-provider-aws/blob/master/pkg/controller/infrastructure/actuator_delete.go#L97
https://github.com/aws/aws-sdk-go/blob/a4094870dd20c585c64cdce86f196831f48b070f/aws/credentials/credentials.go#L247
I don't think we can improve the error message further.

Regarding 2.:
I opened https://github.com/gardener/gardener-extension-provider-aws/pull/209

**Comment by rfranzke** [comment]
The only open point is

> 3. Use more aggressive backoffs in case of rate limit errors

In two discussions we figured that it's not easily achievable because we would probably a) first need a new error code indicating API rate limit issues, and then b) a good classification of those. Especially b) is not trivial.

In favor of other tasks with a higher priority we decided to not follow up for now as and revisit if necessary.

/close
